### PR TITLE
Add Deconvolution2D

### PR DESCRIPTION
**Pull request by t-abe**
#315

I updated @tnarihi's deconvolution implementation to make it working on current chainer, and added tests.

### CODE REVIEW

File: chainer/functions/connection/deconvolution_2d.py [hunk 1]
```diff
@@ -0,0 +1,300 @@
...
+        type_check.expect(in_types.size() == 1)
+        x_type, = in_types
+
+        type_check.expect(
+            x_type.dtype == numpy.float32,
+            x_type.ndim == 4,
+            x_type.shape[1] == self.in_channels
+        )
+
+    def check_type_backward(self, in_types, out_types):
```
Comment (delta2323) [inline chainer/functions/connection/deconvolution_2d.py]: `check_type_backward` is no longer used. (cf. #209)

File: chainer/functions/connection/deconvolution_2d.py [hunk 2]
```diff
@@ -0,0 +1,300 @@
...
+            return 'gW',
+        return 'gW', 'gb'
+
+    def zero_grads(self):
+        self.gW.fill(0)
+        if self.gb is not None:
+            self.gb.fill(0)
+
+    def forward_cpu(self, x):
+        n, c, h, w = x[0].shape
```
Comment (delta2323) [inline chainer/functions/connection/deconvolution_2d.py]: `n, c` is not used. Use `_` instead.

File: chainer/functions/connection/deconvolution_2d.py [hunk 3]
```diff
@@ -0,0 +1,300 @@
...
+        gcol = numpy.tensordot(self.W, x[0], (0, 1))
+        # k, m, n, b, h, w
```
Comment (delta2323) [inline chainer/functions/connection/deconvolution_2d.py]: I would like to add information that `k`, `m`, `n` indicate the # of output channels(`out_channel`), the height of kernels `kh` and the width of kernels `kw`, respectively.

File: chainer/functions/connection/deconvolution_2d.py [hunk 4]
```diff
@@ -0,0 +1,300 @@
...
+    _fwd_pref = libcudnn.CUDNN_CONVOLUTION_FWD_SPECIFY_WORKSPACE_LIMIT
+
+
+def _pair(x):
+    if hasattr(x, '__getitem__'):
+        return x
+    return (x, x)
+
+
+def get_deconv_outsize(h, kh, sy, ph):
```
Comment (delta2323) [inline chainer/functions/connection/deconvolution_2d.py]: I think `get_deconv_outsize` should be in `conv` module in parallel with `get_conv_outsize` .

File: tests/chainer_tests/functions_tests/connection_tests/test_deconvolution_2d.py [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+            -1, 1, self.func.b.shape).astype(numpy.float32)
+        self.func.gW.fill(0)
+        self.func.gb.fill(0)
+
+        self.gy = numpy.random.uniform(-1, 1,
+                                       (2, 2, 5, 3)).astype(numpy.float32)
+        self.x = numpy.random.uniform(-1, 1,
+                                      (2, 3, 3, 2)).astype(numpy.float32)
+
+    def check_forward_consistency(self):
```
Comment (delta2323) [inline tests/chainer_tests/functions_tests/connection_tests/test_deconvolution_2d.py]: Could you add the test case where `nobias` is `True` ?

File: tests/chainer_tests/functions_tests/connection_tests/test_deconvolution_2d.py [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+        self.check_forward_consistency()
+
+    @attr.gpu
+    @condition.retry(3)
+    def test_forward_consistency_im2col(self):
+        self.func.use_cudnn = False
+        self.check_forward_consistency()
+
+    def check_backward(self, x_data, y_grad):
+        x = chainer.Variable(x_data)
```
Comment (delta2323) [inline tests/chainer_tests/functions_tests/connection_tests/test_deconvolution_2d.py]: Could you add test cases where `nobias` is `True` ?

File: tests/chainer_tests/functions_tests/connection_tests/test_deconvolution_2d.py [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+import chainer
+from chainer import cuda
+from chainer import functions as F
+from chainer import gradient_check
+from chainer import testing
+from chainer.testing import attr
+from chainer.testing import condition
+
+
+class TestDeconvolution2D(unittest.TestCase):
```
Comment (delta2323) [inline tests/chainer_tests/functions_tests/connection_tests/test_deconvolution_2d.py]: I would like you to make this test case parametrized with respect to input size, kernel size, stride and pad size, since one of the most difficult parts of (deconvolution) is consistency of input size and output size in edge cases.

Comment (delta2323) [inline tests/chainer_tests/functions_tests/connection_tests/test_deconvolution_2d.py]: You can see the typical usage of parameterized tests [here](https://github.com/pfnet/chainer/blob/6788a84caab30bf8860a5c8e5ba5355ba97bc1d4/tests/cupy_tests/test_ndarray.py#L29)

File: chainer/functions/connection/deconvolution_2d.py [hunk 8]
```diff
@@ -0,0 +1,300 @@
...
+
+    The filter weight has four dimensions :math:`(c_I, c_O, k_H, k_W)`
+    which indicate the number of the number of input channels, output channels,
+    height and width of the kernels, respectively.
+    The filter weight is initialized with i.i.d. Gaussian random samples, each
+    of which has zero mean and deviation :math:`\sqrt{1/(c_I k_H k_W)}` by
+    default. The deviation is scaled by ``wscale`` if specified.
+
+    The bias vector is of size :math:`c_O`.
+    Each element of it is initialized by ``bias`` argument.
```
Comment (delta2323) [inline chainer/functions/connection/deconvolution_2d.py]: Its elements are ...

File: chainer/functions/connection/deconvolution_2d.py [hunk 9]
```diff
@@ -0,0 +1,300 @@
...
+    If ``nobias`` argument is set to True, then this function does not hold
+    the bias parameter.
+
+    Let :math:`X` be the input tensor of dimensions :math:`(n, c_I, h, w)`,
+    :math:`(s_Y, s_X)` be the stride of filter application, and
```
Comment (delta2323) [inline chainer/functions/connection/deconvolution_2d.py]: Remove *be* or insert *be* after `(p_H, p_W)` .

**Comment by delta2323** [comment]
@t-abe  I added comments. Could you check them?

**Comment by t-abe** [comment]
Thank you for reviewing.
I am checking comments.
